### PR TITLE
Support inlay hints

### PR DESCRIPTION
**Pull request by 333fred**
I've based the O# api on the proposed inlay hint LSP api, https://github.com/microsoft/vscode-languageserver-node/blob/main/protocol/src/common/protocol.inlayHint.ts, which is pretty much stable now and 3.17 should be out shortly with this shape. I've also done a couple of refactors while touching code:

* Update to C# 10.
* Add some additional assert utils from roslyn, to make test results more readable and actionable.
* Add an interpolated string handler for logging, so that interpolated strings can be used in logging without a perf penalty when they're uneeded.

### CODE REVIEW

File: src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public async Task<InlayHint> Handle(InlayHintResolveRequest request)
+    {
+        if (!_cache.TryGetFromCache(request.Hint, out var roslynHint, out var document))
+        {
+            return request.Hint;
+        }
+
+        var descriptionTags = await roslynHint.GetDescrptionAsync(document, CancellationToken.None);
```
Comment (JoeRobich) [inline src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs:97]: Opened https://github.com/dotnet/roslyn/pull/60217 for this typo

**Review by JoeRobich — APPROVED** [review]
Looks good to me

**Review by filipw — APPROVED** [review]
very cool

**Comment by 333fred** [comment]
I need to adjust one test to account for platform differences (`int` is a `readonly struct` on some platforms, and that affects the ToString() behavior).

**Pull request merged**